source/core_sntp_serializer.c

```diff
@@ -192,144 +188,99 @@ static uint32_t readWordFromNetworkByteOrderMemory( const uint32_t * ptr )
-     * an acceptable value for this utility.
-     */
-
-    /* Check whether an overflow occurs when the server time is ahead of the client time.
-     * Note: This means that the server time is in NTP era 1 (i.e. after 7 Feb 2036) whereas
-     * client time is in NTP era 0.
-     */
-    if( diffIn64Bits < INT32_MIN )
-    {
-        /* Calculate the actual difference in time between the server and client keeping
-         * the different eras in consideration. */
-        uint32_t diff = ( UINT32_MAX - clientTime ) /* Time from client time to end of NTP era 0.*/
-                        + 1U                        /* Time at epoch of NTP era 1, i.e. & Feb 2036 6h:28m:16s UTC */
-                        + serverTime;               /* Time period after NTP era 1. */
-
-        /* Perform 2's complement negation of the absolute difference value to represent
-         * the true time difference between server and client timestamps that are in
-         * different eras. */
-        calculatedTimeDiff = ( int32_t ) diff;
-    }
-
-    /* Check whether overflow occurs when the client time is ahead of the server time.
-     * Note: In this case, the client time would have overflown in NTP era 1 while
-     */
-    else if( diffIn64Bits > INT32_MAX )
+    int64_t eraAdjustedDiff = 0;
+
+    /* Convert the "seconds" part of timestamps to signed 64 bit integer along with determining
+     * relative NTP era presence of server time relative to client time. */
+    int64_t serverTime = ( int64_t ) serverTimeSec;
+    int64_t clientTime = ( int64_t ) clientTimeSec;
+
+    /* Determine if server time belongs to an NTP era different than the server, and accordingly
+     * update the 64 bit representation to account for the era. */
+    int64_t diffWithNoEraAdjustment = serverTime - clientTime;
+    int64_t diffWithServerEraAdjustment = serverTime + TOTAL_SECONDS_IN_NTP_ERA -
+                                          clientTime;                                /* This helps determine whether server is an
+                                                                                      * era ahead of client time. */
+    int64_t diffWithClientEraAdjustment = serverTime -
+                                          ( TOTAL_SECONDS_IN_NTP_ERA + clientTime ); /* This helps determine whether server is an
+                                                                                      * era ahead of client time. */
+
+
+    /* If the difference value is INT32_MIN, it means that the server and client times are away by
+     *  exactly half the range of second values represented in 32 bits. As a signed 32 bit integer
+     *  cannot represent value of 2^31 as a positive value but it can store it as a negative value,
+     *  this library will treat this special case as the server being behind the client system
+     *  by 2^31 bits. */
+    if( diffWithNoEraAdjustment == INT32_MIN )
```

Comment:
Given the current time is ~3.8b seconds since 1/1/1900, then this can only be true if clientTime is almost 6b seconds, but that value is greater than UINT32_MAX.  Does this case come into play after the server enters NTP era 1?

Reply:
Yes, this case comes into play for any combination of server and client times being exactly `absolute( INT32_MAX ) = 2147483648` seconds apart, so it covers the case when the server is in 2037 but the client is exactly `2147483648` seconds behind sometime in 1969.  With this logic, even if the server is actually ahead of the client time, the calculated clock-offset will be `-2147483648` as the `int32_t` type cannot store the value of `+2147483648`. 

Should we return `ClockOffsetOverflow` instead in that case because the library cannot be certain then whether the client is ahead of the server or vice-versa?

Reply:
Would we be better off with the assumption that the server is ahead in this case, and return INT32_MAX?  The missing second would show up as an offset in the next iteration, yes?

Reply:
We could use that approach, though it would leave the client still behind by a second and be problematic for application that have more less than a second of accuracy requirements. However, the desired accuracy will be achieved in the subsequent time request made after that first iteration.